engine/src/main/java/org/pentaho/di/trans/steps/excelinput/ExcelInput.java

```diff
@@ -0,0 +1,861 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+package org.pentaho.di.trans.steps.excelinput;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.TimeZone;
+
+import org.apache.commons.vfs2.FileObject;
+import org.apache.poi.openxml4j.util.ZipSecureFile;
+import org.pentaho.di.core.Const;
+import org.pentaho.di.core.ResultFile;
+import org.pentaho.di.core.RowSet;
+import org.pentaho.di.core.exception.KettleException;
+import org.pentaho.di.core.exception.KettleFileException;
+import org.pentaho.di.core.fileinput.FileInputList;
+import org.pentaho.di.core.playlist.FilePlayListAll;
+import org.pentaho.di.core.playlist.FilePlayListReplay;
+import org.pentaho.di.core.row.RowMeta;
+import org.pentaho.di.core.row.ValueMetaInterface;
+import org.pentaho.di.core.row.value.ValueMetaNumber;
+import org.pentaho.di.core.spreadsheet.KCell;
+import org.pentaho.di.core.spreadsheet.KCellType;
+import org.pentaho.di.core.spreadsheet.KSheet;
+import org.pentaho.di.core.util.EnvUtil;
+import org.pentaho.di.core.util.Utils;
+import org.pentaho.di.core.vfs.KettleVFS;
+import org.pentaho.di.i18n.BaseMessages;
+import org.pentaho.di.trans.Trans;
+import org.pentaho.di.trans.TransMeta;
+import org.pentaho.di.trans.step.BaseStep;
+import org.pentaho.di.trans.step.StepDataInterface;
+import org.pentaho.di.trans.step.StepInterface;
+import org.pentaho.di.trans.step.StepMeta;
+import org.pentaho.di.trans.step.StepMetaInterface;
+import org.pentaho.di.trans.step.errorhandling.CompositeFileErrorHandler;
+import org.pentaho.di.trans.step.errorhandling.FileErrorHandler;
+import org.pentaho.di.trans.step.errorhandling.FileErrorHandlerContentLineNumber;
+import org.pentaho.di.trans.step.errorhandling.FileErrorHandlerMissingFiles;
+
+/**
+ * This class reads data from one or more Microsoft Excel files.
+ *
+ * @author Matt
+ * @author timh
+ * @since 19-NOV-2003
+ */
+public class ExcelInput extends BaseStep implements StepInterface {
+  private static Class<?> PKG = ExcelInputMeta.class; // for i18n purposes, needed by Translator2!!
+
+  private ExcelInputMeta meta;
+
+  private ExcelInputData data;
+
+  public ExcelInput( StepMeta stepMeta, StepDataInterface stepDataInterface, int copyNr, TransMeta transMeta,
+                     Trans trans ) {
+    super( stepMeta, stepDataInterface, copyNr, transMeta, trans );
+    setZipBombConfiguration();
+  }
+
+  /**
+   * Build an empty row based on the meta-data...
+   *
+   * @return
+   */
+
+  private Object[] fillRow( int startcolumn, ExcelInputRow excelInputRow ) throws KettleException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 42. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 242-306 into a separate method.